# Índice

1. [Importamos librerías y configuraciones](#importamos-librerías-y-configuraciones)
2. [Importamos Datos](#importamos-datos)
3. [Comprobamos longitud de los DataFrames](#comprobamos-longitud-de-los-dataframes)
4. [Unión de DataFrames](#unión-de-dataframes)
5. [Sustitución de Valores Binarios (Yes/No, N/Y)](#sustitución-de-valores-binarios-yesno-ny)
6. [Conversión de Tipo de Datos](#conversión-de-tipo-de-datos)
7. [Salvamos Datos](#salvamos-datos)

---
---

## Importamos librerias y configuraciones

In [61]:
%load_ext autoreload
%autoreload 2

import sys
sys.path.append('../')

from config import *

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


---
---

## Importamos Datos

In [62]:
df_emplo_survey_data = pd.read_csv('../../datos/employee_survey_data.csv').reset_index(drop=True)
df_emplo_survey_data.sample()

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance
4322        4323                     3.00             2.00             2.00

In [63]:
df_general_data = pd.read_csv('../../datos/general_data.csv').reset_index(drop=True)
df_general_data.sample()

Age Attrition BusinessTravel       Department  DistanceFromHome  \
3045   33       Yes  Travel_Rarely  Human Resources                28   

      Education   EducationField  EmployeeCount  EmployeeID  Gender  JobLevel  \
3045          2  Human Resources              1        3046  Female         5   

      JobRole MaritalStatus  MonthlyIncome  NumCompaniesWorked Over18  \
3045  Manager        Single          55610                1.00      Y   

      PercentSalaryHike  StandardHours  StockOptionLevel  TotalWorkingYears  \
3045                 21              8                 1               1.00   

      TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
3045                      3               1                        0   

      YearsWithCurrManager  
3045                     0

In [64]:
df_mana_survey_data = pd.read_csv('../../datos/manager_survey_data.csv').reset_index(drop=True)
df_mana_survey_data.sample()

EmployeeID  JobInvolvement  PerformanceRating
1356        1357               3                  4

---
---

## Comprobamos longitud de los DataFrames

#### (comprobamos que los 3dfs tengan el mismo número de filas)

In [65]:
len_dfs = [df_general_data.shape[0], df_emplo_survey_data.shape[0], df_mana_survey_data.shape[0]]
set(len_dfs)

{4410}

#### como podemos ver, todas tienen 4410 filas, ya que si no, al convertir la lista en un set y por defecto, eliminar los valores repetidos, en caso de haber longitudes diferentes, nos saldrían más de un unico valor.

---
---

## Union DataFrames

#### Ahora que hemos visto que tenemos el mismo número de filas, vamos a unir todos los datos en un dataframe mediante el ```EmployeeID``` [también nos hemos asegurado que EmployeeID tiene el mismo nombre en los 3 dfs]

#### cambiamos los nombres de las columnas del df ```df_mana_survey_data``` para que luego sepamos que pertenecen a la opinión del manager sobre ese empleado.

In [66]:
df_mana_survey_data.columns = ['EmployeeID', 'Manager_opinion_JobInvolvement', 'Manager_opinion_PerformanceRating']
df_mana_survey_data.sample()

EmployeeID  Manager_opinion_JobInvolvement  \
2626        2627                               2   

      Manager_opinion_PerformanceRating  
2626                                  3

In [67]:
df = df_emplo_survey_data.merge(df_general_data, on='EmployeeID').merge(df_mana_survey_data, on='EmployeeID')
df.sample()

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  \
762         763                     1.00             4.00             2.00   

     Age Attrition BusinessTravel              Department  DistanceFromHome  \
762   47        No  Travel_Rarely  Research & Development                23   

     Education EducationField  EmployeeCount  Gender  JobLevel  \
762          4  Life Sciences              1  Female         2   

             JobRole MaritalStatus  MonthlyIncome  NumCompaniesWorked Over18  \
762  Sales Executive       Married          62740                7.00      Y   

     PercentSalaryHike  StandardHours  StockOptionLevel  TotalWorkingYears  \
762                 14              8                 2              17.00   

     TrainingTimesLastYear  YearsAtCompany  YearsSinceLastPromotion  \
762                      0               6                        1   

     YearsWithCurrManager  Manager_opinion_JobInvolvement  \
762                     2                               3   

     Manager_opinion_PerformanceRating  
762                                  3

#### comprobamos si tenemos duplicados

In [68]:
df.duplicated().sum()

0

---
---

## Sustitución Valores Binarios (Yes/No, N/Y).

#### Vamos a sustituir los valores (Yes/No, N/Y) por Booleanos True y False.

#### Variables a modificar: 
- Attrition
- Over18

In [69]:
df['Attrition'].unique()

array(['No', 'Yes'], dtype=object)

In [70]:
df['Over18'].unique()

array(['Y'], dtype=object)

In [71]:
df['Over18'].value_counts()

Over18
Y    4410
Name: count, dtype: int64

#### hemos visto que ```Over18``` tiene un 100 de 'Y' como valor [lo cual tiene sentido], por lo que la vamos a eliminar de nuestro df

In [72]:
df.drop(columns=['Over18'], inplace=True)

df['Attrition'] = df['Attrition'].map({'No': False, 'Yes': True})

In [73]:
df.sample(3)

EmployeeID  EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  \
170          171                     1.00             4.00             3.00   
2822        2823                     1.00             1.00             4.00   
3451        3452                     1.00             3.00             3.00   

      Age  Attrition     BusinessTravel              Department  \
170    47      False      Travel_Rarely  Research & Development   
2822   32       True      Travel_Rarely  Research & Development   
3451   26      False  Travel_Frequently  Research & Development   

      DistanceFromHome  Education    EducationField  EmployeeCount  Gender  \
170                  1          3  Technical Degree              1    Male   
2822                 2          4     Life Sciences              1  Female   
3451                 3          3           Medical              1    Male   

      JobLevel                 JobRole MaritalStatus  MonthlyIncome  \
170          5  Manufacturing Director      Divorced         135700   
2822         2       Research Director        Single          41480   
3451         2   Laboratory Technician      Divorced         139640   

      NumCompaniesWorked  PercentSalaryHike  StandardHours  StockOptionLevel  \
170                 5.00                 16              8                 2   
2822                7.00                 15              8                 0   
3451                1.00                 11              8                 2   

      TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
170               20.00                      2               5   
2822              10.00                      0               5   
3451               5.00                      1               5   

      YearsSinceLastPromotion  YearsWithCurrManager  \
170                         0                     4   
2822                        0                     4   
3451                        1                     3   

      Manager_opinion_JobInvolvement  Manager_opinion_PerformanceRating  
170                                2                                  3  
2822                               2                                  3  
3451                               2                                  3

---
---

## Conversion Tipo Datos

#### Vamos a comprobar el tipo de dato que tienen nuestras varibales junto a su valor y en caso de que sea necesario, cambiarlo, eliminar la columna, etc.

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 28 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   EmployeeID                         4410 non-null   int64  
 1   EnvironmentSatisfaction            4385 non-null   float64
 2   JobSatisfaction                    4390 non-null   float64
 3   WorkLifeBalance                    4372 non-null   float64
 4   Age                                4410 non-null   int64  
 5   Attrition                          4410 non-null   bool   
 6   BusinessTravel                     4410 non-null   object 
 7   Department                         4410 non-null   object 
 8   DistanceFromHome                   4410 non-null   int64  
 9   Education                          4410 non-null   int64  
 10  EducationField                     4410 non-null   object 
 11  EmployeeCount                      4410 non-null   int64

In [75]:
df.describe().T

count     mean      std      min      25%  \
EmployeeID                        4410.00  2205.50  1273.20     1.00  1103.25   
EnvironmentSatisfaction           4385.00     2.72     1.09     1.00     2.00   
JobSatisfaction                   4390.00     2.73     1.10     1.00     2.00   
WorkLifeBalance                   4372.00     2.76     0.71     1.00     2.00   
Age                               4410.00    36.92     9.13    18.00    30.00   
DistanceFromHome                  4410.00     9.19     8.11     1.00     2.00   
Education                         4410.00     2.91     1.02     1.00     2.00   
EmployeeCount                     4410.00     1.00     0.00     1.00     1.00   
JobLevel                          4410.00     2.06     1.11     1.00     1.00   
MonthlyIncome                     4410.00 65029.31 47068.89 10090.00 29110.00   
NumCompaniesWorked                4391.00     2.69     2.50     0.00     1.00   
PercentSalaryHike                 4410.00    15.21     3.66    11.00    12.00   
StandardHours                     4410.00     8.00     0.00     8.00     8.00   
StockOptionLevel                  4410.00     0.79     0.85     0.00     0.00   
TotalWorkingYears                 4401.00    11.28     7.78     0.00     6.00   
TrainingTimesLastYear             4410.00     2.80     1.29     0.00     2.00   
YearsAtCompany                    4410.00     7.01     6.13     0.00     3.00   
YearsSinceLastPromotion           4410.00     2.19     3.22     0.00     0.00   
YearsWithCurrManager              4410.00     4.12     3.57     0.00     2.00   
Manager_opinion_JobInvolvement    4410.00     2.73     0.71     1.00     2.00   
Manager_opinion_PerformanceRating 4410.00     3.15     0.36     3.00     3.00   

                                       50%      75%       max  
EmployeeID                         2205.50  3307.75   4410.00  
EnvironmentSatisfaction               3.00     4.00      4.00  
JobSatisfaction                       3.00     4.00      4.00  
WorkLifeBalance                       3.00     3.00      4.00  
Age                                  36.00    43.00     60.00  
DistanceFromHome                      7.00    14.00     29.00  
Education                             3.00     4.00      5.00  
EmployeeCount                         1.00     1.00      1.00  
JobLevel                              2.00     3.00      5.00  
MonthlyIncome                     49190.00 83800.00 199990.00  
NumCompaniesWorked                    2.00     4.00      9.00  
PercentSalaryHike                    14.00    18.00     25.00  
StandardHours                         8.00     8.00      8.00  
StockOptionLevel                      1.00     1.00      3.00  
TotalWorkingYears                    10.00    15.00     40.00  
TrainingTimesLastYear                 3.00     3.00      6.00  
YearsAtCompany                        5.00     9.00     40.00  
YearsSinceLastPromotion               1.00     3.00     15.00  
YearsWithCurrManager                  3.00     7.00     17.00  
Manager_opinion_JobInvolvement        3.00     3.00      4.00  
Manager_opinion_PerformanceRating     3.00     3.00      4.00

#### **Cambios que vamos a hacer:**

##### **Eliminación de columnas**

- ```EmployeeID``` ya que para entrenar a nuestro modelo predictivo no nos aporta información. Más adelante para otro tipo de estudio si nos puede ser útil.

- ```EmployeeCount``` ya que todos los registros contienen el valor de 1.

- ```StandardHours``` ya que todos los registros contienen el valor de 8 horas.

guardamos los datos para otro posible estudio ya que lo tenemos todo unido y con los tipos de datos correctos.

In [76]:
df.to_pickle('../../datos/futuro_estudio/df_employee_data.pkl')

Procedemos con la eliminación de columnas

In [77]:
df.drop(columns=['EmployeeID', 'EmployeeCount', 'StandardHours'], inplace=True)
df.sample()

EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  Age  \
2551                     4.00             2.00             3.00   22   

      Attrition BusinessTravel              Department  DistanceFromHome  \
2551       True  Travel_Rarely  Research & Development                16   

      Education EducationField  Gender  JobLevel                JobRole  \
2551          2  Life Sciences  Female         2  Laboratory Technician   

     MaritalStatus  MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  \
2551        Single          35790                1.00                 19   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
2551                 3               1.00                      3   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  \
2551               1                        0                     0   

      Manager_opinion_JobInvolvement  Manager_opinion_PerformanceRating  
2551                               3                                  3

#### **Cambio de Valores (categorización de las variables numericas)**

- ```Education```
- ```EnvironmentSatisfaction```
- ```JobSatisfaction```
- ```WorkLifeBalance```
- ```Manager_opinion_JobInvolvement```
- ```Manager_opinion_PerformanceRating```

In [78]:
df['Education'] = df['Education'].map({
                                        1: 'Below College',
                                        2: 'College',
                                        3: 'Bachelor',
                                        4: 'Master',
                                        5: 'Doctor'
                                    })

df['EnvironmentSatisfaction'] = df['EnvironmentSatisfaction'].map({
                                        1: 'Low',
                                        2: 'Medium',
                                        3: 'High',
                                        4: 'Very High'
                                    })

df['JobSatisfaction'] = df['JobSatisfaction'].map({
                                        1: 'Low',
                                        2: 'Medium',
                                        3: 'High',
                                        4: 'Very High'
                                    })

df['WorkLifeBalance'] = df['WorkLifeBalance'].map({
                                        1: 'Bad',
                                        2: 'Good',
                                        3: 'Better',
                                        4: 'Best'
                                    })

df['Manager_opinion_JobInvolvement'] = df['Manager_opinion_JobInvolvement'].map({
                                        1: 'Low',
                                        2: 'Medium',
                                        3: 'High',
                                        4: 'Very High'
                                    })

df['Manager_opinion_PerformanceRating'] = df['Manager_opinion_PerformanceRating'].map({
                                        1: 'Low',
                                        2: 'Good',
                                        3: 'Excellent',
                                        4: 'Outstanding'
                                    })

print("NUESTROS DATOS AHORA")
print("--"*10 + "\n")
df.sample(5)

NUESTROS DATOS AHORA
--------------------



EnvironmentSatisfaction JobSatisfaction WorkLifeBalance  Age  Attrition  \
1472                  Medium          Medium             Bad   32      False   
3514                    High       Very High          Better   37      False   
3180                     Low            High          Better   22      False   
2397                    High            High          Better   43      False   
353                     High            High            Best   35      False   

         BusinessTravel              Department  DistanceFromHome Education  \
1472  Travel_Frequently  Research & Development                17    Master   
3514      Travel_Rarely  Research & Development                23  Bachelor   
3180      Travel_Rarely  Research & Development                 2   College   
2397      Travel_Rarely  Research & Development                 2    Master   
353       Travel_Rarely  Research & Development                 1  Bachelor   

     EducationField  Gender  JobLevel                 JobRole MaritalStatus  \
1472          Other    Male         4         Sales Executive       Married   
3514  Life Sciences    Male         3  Manufacturing Director      Divorced   
3180        Medical  Female         2   Laboratory Technician       Married   
2397  Life Sciences  Female         1      Research Scientist       Married   
353         Medical    Male         1       Research Director        Single   

      MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  StockOptionLevel  \
1472         193280                1.00                 15                 3   
3514         166590                7.00                 16                 1   
3180         195450                0.00                 14                 0   
2397          24220                1.00                 11                 1   
353          108550                3.00                 13                 1   

      TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
1472               5.00                      2               5   
3514               9.00                      2               6   
3180               3.00                      3               2   
2397              14.00                      2              14   
353               17.00                      2               8   

      YearsSinceLastPromotion  YearsWithCurrManager  \
1472                        0                     3   
3514                        1                     3   
3180                        2                     2   
2397                        6                    11   
353                         1                     6   

     Manager_opinion_JobInvolvement Manager_opinion_PerformanceRating  
1472                           High                         Excellent  
3514                           High                         Excellent  
3180                         Medium                         Excellent  
2397                           High                         Excellent  
353                            High                         Excellent

---
---

## Salvamos Datos

Una vez realizada una exploración inicial incluso, una conversión de los datos, guardamos estos y continuaremos realizando nuestro EDA en el notebook ```2_EDA.ipynb```.

In [79]:
df.to_pickle('../../datos/tratados/df_employee_data.pkl')